### PR TITLE
add: support for tilde in bookmarks

### LINKED ISSUE #3032: [FR]: pls support environment variables like `~` or `$HOME` in `bookmarks`
#### Steps to reproduce
1. put sth like `d:~/Documents` in `~/.local/share/ranger/bookmarks`
2. open ranger
3. the recent path gets prepended to `~/Documents`
4. bookmarks file is broken/can't be used

### PR DESCRIPTION
**Pull request by prashantrahul141**
#### ISSUE TYPE
- Improvement/feature implementation

#### RUNTIME ENVIRONMENT
- Operating system and version:  popOS 22.04
- Terminal emulator and version:  alacritty 0.14.0-dev (fd1a3cc7)
- Python version: 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0]
- Ranger version/commit:  ranger-master
- Locale: en_US.UTF-8

#### CHECKLIST
- [x] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [x] All changes follow the code style **[REQUIRED]**
- [x] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [ ] Config files have been updated
- [ ] Changes require documentation to be updated
    - [ ] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### DESCRIPTION
While loading saved bookmarks, we check if the path starts with `~` if it does then we replace the first occurrence of `~` with the current home directory.

#### MOTIVATION AND CONTEXT
This fixes #3032 

#### TESTING
Manual tests were conducted with various edge cases to ensure the changes doesn't break anything.

### CODE REVIEW

**Review by toonn — CHANGES_REQUESTED** [review]
Use `os.path.expanduser` instead.

**Comment by prashantrahul141** [comment]
I had previously noticed that the bookmarks file gets overwritten with an absolute path after opening Ranger, even if you have something like `~/Documents` in it. Initially, I thought this wasn't really a problem. However, I later realized that this would completely undermine the option of syncing bookmarks if you have a different username, since the absolute path includes the username as well.

So now `FileSystemObject` also keeps a reference to the original path it was given before expanding it to the absolute path. When saving bookmarks, it ensures that the value is of type `FileSystemObject` and use `original_path`, otherwise the old behaviour.

File: ranger/container/bookmarks.py [hunk 1]
```diff
@@ -183,7 +182,7 @@ def save(self):
                 for key, value in self.dct.items():
                     if key in ALLOWED_KEYS \
                             and key not in self.nonpersistent_bookmarks:
-                        key_value = "{0}:{1}\n".format(key, value)
+                        key_value = "{0}:{1}\n".format(key, value.original_path)
```
Comment (toonn) [inline ranger/container/bookmarks.py]: If a bookmark is updated this value is updated too?

**Review by toonn — COMMENTED** [review]
Most of it looks OK.

File: ranger/container/bookmarks.py (hunk 1 above)
Comment (prashantrahul141) [reply ranger/container/bookmarks.py]: https://github.com/user-attachments/assets/f71179e5-186a-4aaa-8698-46c79d4fbffc



is this what you mean?

**Comment by EmmAanue** [comment]
EXe

**Review by EmmAanue — APPROVED** [review]

File: ranger/container/bookmarks.py (hunk 1 above)
Comment (toonn) [reply ranger/container/bookmarks.py]: Yep, perfect.

**Review by toonn — DISMISSED** [review]
LGTM

**Comment by toonn** [comment]
Can you clean up the commit history a bit and rebase?

**Comment by prashantrahul141** [comment]
:sweat_smile: my bad

**Review by toonn — APPROVED** [review]
LGTM

For future reference, multiple commits in a PR are fine, it's just things that amount to reverts and fixups that are better cleaned up before merging.
